src/compiler/emitter.ts

```diff
@@ -3826,78 +3883,158 @@ module ts {
                         emitExpressionForPropertyName(member.name);
                         emitEnd(member.name);
                         write("))");
+                        decreaseIndent();
                     }
 
                     write(");");
                     emitEnd(member);
                     writeLine();
-                });
+                }
             }
-            
-            function emitDecoratorsOfParameters(node: ClassLikeDeclaration, member: FunctionLikeDeclaration) {
-                forEach(member.parameters, (parameter, parameterIndex) => {
-                    if (!nodeIsDecorated(parameter)) {
-                        return;
+
+            function emitDecoratorsOfParameters(node: FunctionLikeDeclaration, writeComma: boolean): boolean {
+                if (node) {
+                    let parameterIndex = 0;
+                    for (let parameter of node.parameters) {
+                        if (nodeIsDecorated(parameter)) {
+                            writeComma = emitDecoratorArray(parameter.decorators, writeComma, `__param(${parameterIndex}, `, ")");
+                        }
+                        ++parameterIndex;
                     }
+                }
+                return writeComma;
+            }
 
-                    // Emit the decorators for a parameter. Given the following:
-                    //
-                    //   class C {
-                    //     constructor(@dec p) { }
-                    //     method(@dec p) { }
-                    //     set accessor(@dec value) { }
-                    //   }
-                    //
-                    // The emit for a constructor is:
-                    //
-                    //   __decorate([dec], C, void 0, 0);
-                    //
-                    // The emit for a parameter is:
-                    //
-                    //   __decorate([dec], C.prototype, "method", 0);
-                    //
-                    // The emit for an accessor is:
-                    //
-                    //   __decorate([dec], C.prototype, "accessor", 0);
-                    //
+            function emitDecoratorArray(decorators: NodeArray<Decorator>, writeComma: boolean, prefix?: string, postfix?: string): boolean {
+                if (decorators) {
+                    let decoratorCount = decorators ? decorators.length : 0;
+                    for (let i = 0; i < decoratorCount; i++) {

```

Comment:
emitList emits the provided node array. I don't want to emit the decorators array, I want to emit the expression of each decorator in the array with the correct source maps. I also don't want to have an emitDecorator that just emits the expression of the decorator, so that if/when we add an ES7 or edge script target in the future I would be using emitDecorator to emit the decorator as is.